[AN-Issue-2002] Updated automation registry to support governance submitted tasks

aregng opened the pull request:
- Introduced registry configuration and state for system tasks
- Created a new version of the active registry configduration to enclose configurations for both user and governance submitted transactions
 - Introduce a new API for system task registration, cancellation
 - Updated task processing to handle avoid charging and refunds for system tasks
 - Updated already registered task-metadata to mark them as user-tasks during migration
 - Introduced view API to access system task data
 - Introduced API to grant/revoke authorization to/from accounts for system task registration

Commit:
[AN-Issue-2002] Updated automation registry to support governance submitted tasks

- Introduced registry configuration and state for system tasks
- Created a new version of the active registry configuration to enclose configurations for both user and governance submitted transactions

TODO:
 - Introduce a new API for system task registration, cancellation
 - Update task processing to handle avoid charging and refunds for system tasks
 - Update already registered task-metadata to mark them as user-tasks during migration
 - Introduce view API to access system task data
 - Introduce API to authorize/deauthorize accounts for system task
   registration

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1115,6 +1181,12 @@ module supra_framework::automation_registry {
+        migrate_registry_config(supra_framework, sys_task_duration_cap_in_secs,  sys_registry_max_gas_cap, sys_task_capacity);
+
+        // Initialize registry state for system tasks
+        initialize_registry_system_task_state(supra_framework);
```
shouldn't this `initialize_registry_system_task_state` also take the extra three parameters being passed to `migrate_v2`? Otherwise the parameters are not utilized anywhere. @aregng

aregng replied:
Dr. @sjoshisupra 
2 new data types have been introduced:
- one for the static configurations for system tasks (`GST`) which are updated by governance: `RegistryConfigForSystemTasks`
   - These parameters needs to be initialized with the agreed parameters during migration
   - a new version of active configuration parameters has been introduced which will enclose now already existing registry configuration parameters for the `UST` and also new configuration parameters for `GST`
- one for the dynamic configurations/state which are updated at runtime, when tasks are registered: `RegistryStateForSystemTasks`
   - These parameters will be updated at runtime, and are set to empty/0 values during initialization.
   - Although in this version of the PR I have `RegsitryStateForSystemTasks` has been initialized as a separate resource, but I am thinking to have a single resource of the registry state `AutomationRegistryV2` which will be
   ```
   main_state: AutomationRegistry,
   system_task_state: RegistryStateForSystemTasks
   ```
Please let me know of your feedback on this approach.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -2403,6 +2507,45 @@ module supra_framework::automation_registry {
+    fun initialize_registry_system_task_state(supra_framework: &signer) {
+        system_addresses::assert_supra_framework(supra_framework);
+        move_to(supra_framework, RegistryStateForSystemTasks {
+            gas_committed_for_this_cycle: 0,
+            gas_committed_for_next_cycle: 0,
+            authorized_accounts: vector[],
+            task_ids: vector[],
```
I think all of this should come as parameters. @aregng

aregng replied:
Please see my reply  here: https://github.com/Entropy-Foundation/aptos-core/pull/276#discussion_r2297466982

Commit:
Updated automation registry with new API for system task registration

- Introduce a new API for system task registration, cancellation
- Update task processing to handle avoid charging and refunds for system tasks
- Update already registered task-metadata to mark them as user-tasks during migration

Commit:
Added view APIs for system task data

Commit:
Update automation concpets in rust layer according to move layer changes

Commit:
Added unit tests

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -58,8 +60,8 @@ module supra_framework::automation_registry {
     const EGAS_COMMITTEED_VALUE_UNDERFLOW: u64 = 13;
-    /// Auxiliary data during registration is not supported
-    const ENO_AUX_DATA_SUPPORTED: u64 = 14;
+    /// Invalid number of auxaliry data.
```
```suggestion
    /// Invalid number of auxilary data.
```

aregng replied:
Done

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -104,6 +106,26 @@ module supra_framework::automation_registry {
+    /// Automation registry max gas capacity for system tasks cannot be zero.
+    const EREGISTRY_MAX_GAS_CAP_NON_ZERO_SYS: u64 = 37;
+    /// Current automation cycle interval is greater than specified system task duration cap.
+    const EUNACCEPTABLE_SYS_TASK_DURATION_CAP: u64 = 38;
```
```suggestion
    const EUNACCEPTABLE_SYS_TASK_DURATION: u64 = 38;
```

aregng replied:
This constant is used to report an error when `system_task_duration_cap` is being updated, and it is checked against `cycle_duration`.
Please let me know whether the `CAP` should be still removed.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -125,6 +147,10 @@ module supra_framework::automation_registry {
     const CANCELLED: u8 = 2;
 
+    /// Constants decribing the task type, USER SUBMITTED TASK (UST - 1), GOVERNANCE SUBMITTED TASK(GST - 0)
+    const GST: u8 = 0;
+    const UST: u8 = 1;
```
Just a thought, should we have only non-zero values as valid type? That way zero can perhaps indicate unassigned/uninitialized/default value? @aregng

aregng replied:
I think as long as we cannot have a task that does not have a specific type, there is no need to have a variant representing that state. And as long as we have 2 separate APIs to register each type of the task, then I think the default value is also redundant.
As for the discriminant we can start from 1.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -171,14 +197,29 @@ module supra_framework::automation_registry {
+        next_cycle_registry_max_gas_cap: u64,
+        /// Flag indicating whether the task registration is enabled or paused.
+        /// If paused a new task registration will fail.
+        registration_enabled: bool,
```
Is this field only for GST or does this flag control all registration? @aregng

sjoshisupra replied:
should we add extra field for future expansion of registry fields? @aregng

aregng replied:
The flag is to control all type of registrations.
Yes we can add an extra field for future expansions.

aregng replied:
Done

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -197,6 +238,20 @@ module supra_framework::automation_registry {
+        /// Exceeding this limit during task registration will cause failure and is used in fee calculation.
+        registry_max_gas_cap: u64,
+        /// Maximum number of system tasks that registry can hold.
+        task_capacity: u16,
```
Same here, should we add extra field for future expansion? @aregng

aregng replied:
Done.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -245,6 +308,26 @@ module supra_framework::automation_registry {
+        /// Total committed max gas amount at the beginning of the current cycle.
+        gas_committed_for_this_cycle: u64,
+        /// Cached system task indexes
+        task_ids: vector<u64>,
```
is `task_ids` going to be a subset of `epoch_active_task_ids`?

aregng replied:
yes, it will contain all system task indexes available in the registry.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -245,6 +308,26 @@ module supra_framework::automation_registry {
     }
 
+    /// It tracks entries both pending and completed, organized by unique indices.
+    struct RegistryStateForSystemTasks has key, store {
```
This should not require `key` ability of it's going to be a field of `AutomationRegustryV2`

aregng replied:
Done

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -197,6 +238,20 @@ module supra_framework::automation_registry {
+    #[resource_group_member(group = supra_framework::object::ObjectGroup)]
+    #[event]
+    /// Automation registry configuration parameters for governance/system submitted tasks
+    struct RegistryConfigForSystemTasks has key, store, drop, copy {
```
This may not require `key` ability if it's going to be a field of `ActiveAutomationRegistryConfigV2`

aregng replied:
Done

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -614,48 +736,48 @@ module supra_framework::automation_registry {
-        let automation_task_metadata = borrow_global<AutomationRegistry>(@supra_framework);
-        if (enumerable_map::contains(&automation_task_metadata.tasks, task_index)) {
-            let value = enumerable_map::get_value_ref(&automation_task_metadata.tasks, task_index);
+    public fun has_sender_active_task_with_id(sender: address, task_index: u64): bool acquires AutomationRegistryV2 {
```
@aregng, there was an additional requirement that even if the task is not active, one should be able to query if a particular `address` is the `owner` of the task, this was needed for AutoFi (CC: @nizam-supraoracles) I do not recall if this was already added.

aregng replied:
Dr. @sjoshisupra , it was added and is part of the dev branch. I have not yet rebased the feature branch to the latest one that's why it is missing from this version.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -862,33 +1021,64 @@ module supra_framework::automation_registry {
+        if (vector::contains(&system_tasks_state.authorized_accounts, &account)) {
+            return
+        };
+        // TODO: Clarify, should the account be checked to be an existing multisig account ?
```
@aregng, yes for now, it makes sense to add the validation that this is indeed a multisig account being added/authorized.

aregng replied:
Done

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1064,7 +1256,126 @@ module supra_framework::automation_registry {
+        vector::for_each(task_indexes, |task_index| {
+            if (enumerable_map::contains(&automation_registry.main.tasks, task_index)) {
+                // Remove task from registry
+                let task = enumerable_map::remove_value(&mut automation_registry.main.tasks, task_index);
```
We also need to remove this from cached `task_ids` from system registry, right? @aregng

aregng replied:
Sure it is done at line 1293 below.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1064,7 +1256,126 @@ module supra_framework::automation_registry {
+    /// Committed gas-limit is updated by reducing it with the max-gas-amount of the cancelled task.
+    public entry fun cancel_system_task(
+        owner_signer: &signer,
+        task_index: u64
```
I am wondering if we should make this parameter a vector too, just like in the stopping API @aregng

aregng replied:
we can, I just kept it similar to the UST api.
Let me add it in a separate PR.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1064,7 +1256,126 @@ module supra_framework::automation_registry {
+        assert!(enumerable_map::contains(&automation_registry.main.tasks, task_index), EAUTOMATION_TASK_NOT_FOUND);
+
+        let automation_task_metadata = enumerable_map::get_value(&mut automation_registry.main.tasks, task_index);
+        assert!(is_of_type(&automation_task_metadata, GST), EUNSUPPORTED_TASK_OPERATION);
```
Isn't this check redundant if we are already checking at line 1344 if `task_index` is part of `automation_regustry.system_tasks_state.task_ids`? @aregng

aregng replied:
it is, unless there is a regression in the implementation.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1232,20 +1577,20 @@ module supra_framework::automation_registry {
     /// If native automation feature is enabled and automation lifecycle has been in CYCLE_READY state,
     /// then lifecycle is restarted.
-    public(friend) fun on_new_epoch() acquires AutomationCycleDetails, ActiveAutomationRegistryConfig, AutomationRegistry {
+    public(friend) fun on_new_epoch() acquires AutomationCycleDetails, ActiveAutomationRegistryConfigV2, AutomationRegistryV2 {
```
When will this be called now? We don't need this to be called at epoch boundary block , right? @aregng

aregng replied:
Right, It is called when chain-state is reconfigured.
https://github.com/Entropy-Foundation/aptos-core/blob/db54c00e5645506a9a7e6a5a544ce677532e8146/aptos-move/framework/supra-framework/sources/reconfiguration_with_dkg.move#L58

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1368,13 +1714,90 @@ module supra_framework::automation_registry {
+            gas_price_cap: 0,
+            // No Automation fee cap, as system tasks are not charged
+            automation_fee_cap_for_epoch: 0,
+            aux_data,
```
Where are we initializing `aux_data`? This should be initialized to `GST`.

aregng replied:
It is expected to be initialized in native layer, when registration transaction is created.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -1971,7 +2470,8 @@ module supra_framework::automation_registry {
         vector::for_each(ids, |task_index| {
-            let task = enumerable_map::get_value_ref(&automation_registry.tasks, task_index);
+            let task = enumerable_map::get_value_mut(&mut automation_registry.tasks, task_index);
+            task.aux_data = vector[vector[UST]];
```
This can create problems. `ids` is getting the list from `automation_registry.tasks` which also contains sytem tasks but now this would set `aux_data` of all tasks to `UST`. Also, I am not even sure if processing of entire registry in Move layer, when there are thousands of tasks would be feasible. @aregng

aregng replied:
This function is utilized only during migration to V2.
So it is expected that all the tasks are only user registered.
As for being able to traverse all the registered tasks, I guess it should be possible as the maximum tasks that are expected to be traversed should be no more than 500 (max # of tasks supported in V1) and we should not cross any limits, as 500 task traversal and charging has been tested for epoch-changes in scope of block metadata transaction execution, and please note that the migration will be run as governance action and gov-txns have higher limits.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -2338,71 +2839,139 @@ module supra_framework::automation_registry {
+        let type_value = vector::borrow(maybe_task_type, 0);
+        assert!(*type_value == task_type, EINVALID_TASK_TYPE)
+    }
+
+    fun migrate_registry_config(
```
```suggestion
    // Precondition: `ActiveAutomationRegistryConfig` must exist at `@supra_framework`
    // Postcondition: `ActiveAutomationRegistryConfig` must not exist at `@supra_framework` AND `ActiveAutomationRegistryConfigV2` must exist
    //                 at `@supra_framework`
    fun migrate_registry_config(
```

aregng replied:
Done

Commit:
Addressed review comments

- Added auxiliary configuration options for easy expansion after release

sjoshisupra reviewed (commented):
Shall we move all the `tests` to a different module `automation_registry_tests.move` under `tests` folder (should be a sibling of `sources`)? This would keep file size low and especially reduce size when we publish with artifacts.

In general, moving forward, we would like all the tests always be in a separate `tests` folder. In source files, minimal number of `test_only` function may be present to expose functionalities to allow them to be tested from a different test module.

aregng commented:
> Shall we move all the `tests` to a different module `automation_registry_tests.move` under `tests` folder (should be a sibling of `sources`)? This would keep file size low and especially reduce size when we publish with artifacts.
> 
> In general, moving forward, we would like all the tests always be in a separate `tests` folder. In source files, minimal number of `test_only` function may be present to expose functionalities to allow them to be tested from a different test module.

If It is acceptable I would like to address this in separate PR.
